core/controllers/acl_decorators_test.py

```diff
@@ -2165,6 +2165,10 @@ def test_assigning_voice_artist_for_unsupported_entity_type_raise_400(self):
         self.logout()
 
     def test_voiceover_admin_can_manage_voice_artist_in_private_exp(self):
+        # Voiceover admin can manage voice artists in private explorations, this
```

Comment:
Some minor grammar rewording:

```
Technically, voiceover admins can manage voice artists in any exploration, including private ones. Note, however, that this does not mean that they can actually assign voice artists to private explorations. The logic of assigning voice artists to different types of explorations is handled in rights_manager.py.
```

But actually on thinking through this I am a bit worried about the philosophy here. It seems to me like the decorator should analyze the action and just reject if it's not an allowed action, otherwise it's kind of confusing to have it like "you can do this ... but then, later on, you can't". Would it be possible to have the rejection occur at this level if a user is trying to do something that they aren't allowed to do?

Reply:
Hmm, I'm not sure I agree. I always understood ACL decorators as stuff that some particular user can do (and some other users cannot) and that it shouldn't be used for setting universal rules that all users cannot do. 

In this case, the universal rule is that we cannot add voice artists to unpublished exploration (the only way would be to modify the datastore directly). That universal rule is encoded in the _rights_manager.py_, which is also verified by tests. Then the per-user rule is whether some particular user has enough rights to add or remove voice artists from exploration. Because of that, I don't think we should check whether the exploration is published in the ACL decorators.

A different case would be if we had a separate permission that would allow adding voice artists to unpublished explorations (for example super admins would be able to do this). In that case, I think it would make sure to check whether the exploration is published even in the ACL decorators.

Reply:
Updated the comment.

Reply:
Hm I think that's reasonable as a general principle, but we should be consistent. What do other handlers do, and in particular do they consider the public/private status of an exploration?

If they don't, I'm fine with this, but if they do, let's be consistent with that (or discuss further).

Reply:
I've gone through the handlers, most of them take the user into account. Some of them take other stuff into account but only when it actually depends on the user. Like super admin can modify any exploration, curriculum admin can review only when the exploration is published. From that I don't think it makes sense to take the published/private into account here. But if you think it should be added then I would rather split this into two access methods one for adding and one for removing the voice artist.

Reply:
Yeah, let's go with the split. I think that would be better and would ensure consistency up and down the chain. Thanks for investigating!